examples/virtual_methods/base_button/imp.rs

```diff
@@ -0,0 +1,115 @@
+use super::BaseButtonExt;
+use gtk::{
+    gio,
+    glib::{self, translate::from_glib_borrow, Error},
+    prelude::*,
+    subclass::prelude::*,
+};
+use std::{future::Future, pin::Pin};
+
+pub type BaseButtonInstance = <BaseButton as super::ObjectSubclass>::Instance;
+pub type PinnedFuture = Pin<Box<dyn Future<Output = Result<(), Error>> + 'static>>;
+
+/// GObject glue code for our BaseButtonClass which holds the function pointers to our virtual functions.
+#[repr(C)]
+pub struct BaseButtonClass {
+    pub parent_class: gtk::ffi::GtkButtonClass,
+    pub sync_method:
+        Option<unsafe extern "C" fn(*mut BaseButtonInstance, extra_text: *mut Option<String>)>,
+    pub async_method: Option<unsafe extern "C" fn(*mut BaseButtonInstance) -> PinnedFuture>,
+}
+
+unsafe impl ClassStruct for BaseButtonClass {
+    type Type = super::imp::BaseButton;
+}
+
+impl std::ops::Deref for BaseButtonClass {
+    type Target = glib::Class<glib::Object>;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { &*(self as *const _ as *const Self::Target) }
+    }
+}
+
+impl std::ops::DerefMut for BaseButtonClass {
+    fn deref_mut(&mut self) -> &mut glib::Class<glib::Object> {
+        unsafe { &mut *(self as *mut _ as *mut glib::Class<glib::Object>) }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct BaseButton {}
+
+// Virtual method default implementation trampolines
+unsafe extern "C" fn sync_method_default_trampoline(
+    this: *mut BaseButtonInstance,
+    extra_text: *mut Option<String>,
+) {
+    let imp = (*this).impl_();
+    imp.sync_method(&from_glib_borrow(this), Box::from_raw(extra_text))
+}
+
+unsafe extern "C" fn async_method_default_trampoline(
+    this: *mut BaseButtonInstance,
+) -> PinnedFuture {
+    let imp = (*this).impl_();
+    imp.async_method(&from_glib_borrow(this))
+}
+
+pub unsafe extern "C" fn base_button_sync_method(
+    this: *mut BaseButtonInstance,
+    extra_text: *mut Option<String>,
+) {
+    let klass = glib::subclass::types::InstanceStruct::class(&*this);
+
+    (klass.sync_method.unwrap())(this, extra_text)
+}
+
+pub unsafe extern "C" fn base_button_async_method(this: *mut BaseButtonInstance) -> PinnedFuture {
+    let klass = glib::subclass::types::InstanceStruct::class(&*this);
+    klass.async_method.unwrap()(this)
+}
+
+/// Default implementations of our sync_method and async_method.
+impl BaseButton {
+    fn sync_method(&self, obj: &super::BaseButton, extra_text: Box<Option<String>>) {
+        if let Some(text) = *extra_text {
+            obj.set_label(&format!("BaseButton sync: {}", text));
+        } else {
+            obj.set_label("BaseButton sync");
+        }
+    }
+
+    fn async_method(&self, obj: &super::BaseButton) -> PinnedFuture {
```

Comment:
This should maybe also take `extra_text` for consistency

Reply:
Hm, I actually thought it was nice to display functions that take arguments vs functions which don't

Reply:
No strong opinion here. @bilelmoussaoui @GuillaumeGomez?

Reply:
No opinion either.

Reply:
Is passing a param in the async case any different ?